src/Processors/Transforms/Streaming/JoinTransformWithAlignment.h

```diff
@@ -27,81 +30,120 @@ class JoinTransformWithAlignment final : public IProcessor
     Status prepare() override;
     void work() override;
 
+    void checkpoint(CheckpointContextPtr ckpt_ctx) override;
+    void recover(CheckpointContextPtr ckpt_ctx) override;
+
     static Block transformHeader(Block header, const HashJoinPtr & join);
 
 private:
-    Status prepareLeftInput();
-    Status prepareRightInput();
-
-    inline Int64 getRightWatermark(const Chunk & chunk) const
+    struct InputPortWithData
     {
-        return getWatermark(chunk, right_watermark_column_position, right_watermark_column_type);
-    }
+        explicit InputPortWithData(InputPort * input_port_) : input_port(input_port_) { }
 
-    inline Int64 getLeftWatermark(const Chunk & chunk) const
-    {
-        return getWatermark(chunk, left_watermark_column_position, left_watermark_column_type);
-    }
+        InputPort * input_port;
```

Comment:
Let's move the data members to the end of the struct. `ctor / dtor, member functions, data mebers`